### ISSUE TITLE
How does retrain work?

### ISSUE THREAD

**Issue by little-eyes** [Microsoft/LightGBM#1469]
Hi,

I was looking at the following code to retrain the model from an existing model.

https://github.com/Microsoft/LightGBM/blob/c803b5cbca7b53ad83b473ffa1e1ed0436015e5c/examples/python-guide/advanced_example.py#L99-L107

Can anyone explain a little bit more or point to the related code that helps understand how does this work? Is that going to update the tree's parameters only or it will add new trees into the booster, or it will update the tree structures from the initial model? Or it will do other magic things?

Thanks,
Jilong

**Comment by guolinke** [comment]
This solution is adding new trees to the `model.txt`

**Comment by little-eyes** [comment]
Thanks @guolinke for your reply!

**Comment by little-eyes** [comment]
Well, how is this doable without having the previous data? 

AFIAK, this process cannot be additive. By adding in additional trees without the previous data, you risk degrade accuracy for data with previous predictive structures. That is, the tree has a fixed structure already, fitted to the structure of the existing data, without having the existing data to evaluate updates (i.e. more trees) you can bias the updated model.

**Comment by guolinke** [comment]
Sorry, I didn't get it.
The previous trees in `init_model` is fixed in continued-train. The new data will use previous trees to have a prediction first, then use the residual to train the new trees. The previous data is not used in this process.

**Comment by little-eyes** [comment]
I see. So if you add more trees into the model after each continued training, will it overfit? For example, if I need to train 5 times, and each time uses the last training's model as initial model, but the preview training data is not available, so I will end up with 5x more trees than I expected. Since the previous training's trees will not be updated, it might overfit?

**Comment by guolinke** [comment]
if the distribution of your data are almost the same and relative large at 5 times, It will not overfit. Otherwise, it may overfit.

**Comment by Quetzalcohuatl** [comment]
Are there any guidelines to how long to retrain for? We originally train using a validation set, but all incremental training has to be done without a validation set. Maybe I can have a rule like this:

```
LENGTH_ORIGINAL_DATA = len(train_data)
ORIGINAL_NUM_ITERATIONS = 1000 # Whatever your best iterations was

LENGTH_INCREMENTAL_DATA = len(new_data) # Presumably smaller than your train data if you want to train incrementally as you receive new data
RETRAIN_NUM_ITERATIONS = (LENGTH_INCREMENTAL_DATA / LENGTH_ORIGINAL_DATA) * ORIGINAL_NUM_ITERATIONS
```

Example: Suppose you train on 1,000,000 rows and then receive 1,000 new rows. If you originally trained for 1,000 iterations, then you will retrain for only 1 iteration.

Meh. Only 1 iteration!? Maybe I should do `max(10, RETRAIN_NUM_ITERATIONS)`. Anyone else come up with a solution for this, or do most people just retrain for only 10 iterations like the OP was recommending?
